### PR TITLE
Eliminate the allocation on each aroundRecieve

### PR DESCRIPTION
**Pull request by johanandren**
Refs #20748

### CODE REVIEW

**Comment by johanandren** [comment]
Credits go to @retronym, thanks!

File: akka-actor/src/main/scala/akka/actor/Actor.scala [hunk 1]
```diff
@@ -481,7 +491,11 @@ trait Actor {
    * @param receive current behavior.
    * @param msg current message.
    */
-  protected[akka] def aroundReceive(receive: Actor.Receive, msg: Any): Unit = receive.applyOrElse(msg, unhandled)
+  protected[akka] def aroundReceive(receive: Actor.Receive, msg: Any): Unit = {
+    if (receive.applyOrElse(msg, Actor.notHandledFun) == Actor.NotHandled) {
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: add a comment above the `if` that we do this to avoid allocations?

**Review by ktoso — APPROVED** [review]
Very nice, LGTM :)

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-actor/src/main/scala/akka/actor/Actor.scala [hunk 2]
```diff
@@ -481,7 +491,12 @@ trait Actor {
    * @param receive current behavior.
    * @param msg current message.
    */
-  protected[akka] def aroundReceive(receive: Actor.Receive, msg: Any): Unit = receive.applyOrElse(msg, unhandled)
+  protected[akka] def aroundReceive(receive: Actor.Receive, msg: Any): Unit = {
+    // optimization: avoid allocation of lambda
+    if (receive.applyOrElse(msg, Actor.notHandledFun) == Actor.NotHandled) {
```
Comment (drewhk) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: eq

Comment (drewhk) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: asInstanceOf[AnyRef] then? I am pretty sure that cast is optimized away as it is meaningless.

File: akka-actor/src/main/scala/akka/actor/Actor.scala [hunk 3]
```diff
@@ -364,6 +364,16 @@ object Actor {
    * that will be translated to the receiving system's deadLetters.
    */
   final val noSender: ActorRef = null
+
+  /**
+   * INTERNAL API
+   */
+  private[akka] final val NotHandled = new Object
```
Comment (retronym) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: Could actually be `private` if only used in the companion.

**Comment by johanandren** [comment]
Unrelated test failure :/

**Comment by johanandren** [comment]
PLS BUILD

**Comment by johanandren** [comment]
PLS BUILD

**Pull request merged**